src/WPGraphQL.php

```diff
@@ -422,50 +424,127 @@ public function init_admin() {
-			$wp_taxonomies['category']->graphql_single_name = 'category';
-			$wp_taxonomies['category']->graphql_plural_name = 'categories';
+		return $args;
+	}
+
+	/**
+	 * This sets the post type /taxonomy GraphQL properties.
+	 *
+	 * @param array  $args      Array of arguments for registering the type.
+	 * @param string $wp_type   Post type / Taxonomy type key.
+	 *
+	 * @return array
+	 */
+	public static function set_wp_type_args( $args, $wp_type ) {
+
+		// Bail early if the post type is hidden from the WPGraphQL schema.
+		if ( empty( $args['show_in_graphql'] ) ) {
+			return $args;
 		}
 
-		// Adds GraphQL support for tags.
-		if ( isset( $wp_taxonomies['post_tag'] ) ) {
-			$wp_taxonomies['post_tag']->show_in_graphql     = true;
-			$wp_taxonomies['post_tag']->graphql_single_name = 'tag';
-			$wp_taxonomies['post_tag']->graphql_plural_name = 'tags';
+		$defaults = [
+			// The "kind" of GraphQL type to register. Can be `interface`, `object`, or `union`.
+			'graphql_kind'                     => 'object',
+			// The callback used to resolve the type. Only used if `graphql_kind` is an `interface` or `union`.
+			'graphql_resolve_type'             => null,
+			// An array of custom interfaces the type should implement.
+			'graphql_interfaces'               => [],
+			// An array of default interfaces the type should exclude.
+			'graphql_exclude_interfaces'       => [],
+			// An array of custom connections the type should implement.
+			'graphql_connections'              => [],
+			// An array of default connection field names the type should exclude.
+			'graphql_exclude_connections'      => [],
+			// An array of possible type the union can resolve to. Only used if `graphql_kind` is a `union`.
+			'graphql_union_types'              => [],
+			// Whether to register default connections to the schema.
+			'graphql_register_root_field'      => true,
+			'graphql_register_root_connection' => true,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
+
+		// Ensure a valid type resolver is set for interface and union types.
+		if ( ! is_callable( $args['graphql_resolve_type'] ) && ( 'interface' === $args['graphql_kind'] || 'union' === $args['graphql_kind'] ) ) {
+			throw new Exception(
+				sprintf(
+					__( '%1$s is registered as a GraphQL %2$s, but has no way to resolve the type. Ensure $args[\'graphql_resolve_type\'] is a valid callback function.', 'wp-graphql' ),
+					$wp_type,
+					$args['graphql_kind']
+				)
+			);
 		}
 
-		// Adds GraphQL support for post formats.
-		if ( isset( $wp_taxonomies['post_format'] ) ) {
-			$wp_taxonomies['post_format']->show_in_graphql     = true;
-			$wp_taxonomies['post_format']->graphql_single_name = 'postFormat';
-			$wp_taxonomies['post_format']->graphql_plural_name = 'postFormats';
+		// Ensure union types have a valid list of possible types.
+		if ( empty( $args['graphql_union_types'] ) && 'union' === $args['graphql_kind'] ) {
+			throw new Exception(
```

Comment:
I think this might be too aggressive. I'm seeing fatal errors when I register a post type like so, then load the admin page: 

```php
add_action( 'init', function() {

	register_post_type( 'testType', [
		'show_in_graphql'     => true,
		'graphql_single_name' => 'TestType',
		'graphql_plural_name' => 'TestTypes',
		'graphql_kind'        => 'union',
	] );

} );
```

I think we should be more graceful with this, and either show an error in GraphQL requests, or skip the registration of the invalid type and show a Debug message. 

In any case, I think it should be limited to GraphQL environment rather than throwing an Exception on every WordPress load. 

![Uploading CleanShot 2022-09-08 at 13.05.32.png…]()

Reply:
If we move the logic to check if both `graphql_kind === 'union' && empty( $post_type_object->graphql_union_types )` to PostObject.php where we call `register_graphql_union_type()` I think we can enable a more graceful experience, which I think will be important for when plugins (Custom Post Type UI, etc) register post types and might allow users to leave fields unpopulated. 

I'd hate to see a full on fatal error for the whole site when it's just a mistake affecting WPGraphQL. 

I think we could do something to the tune of: 

```php
if ( ! empty( $post_type_object->graphql_union_types ) ) {
    
    // Set the possible types for the union.
    $config['typeNames'] = $post_type_object->graphql_union_types;

    register_graphql_union_type( $single_name, $config );

} else {

    graphql_debug( 'Registering a post type with "graphql_kind" => "union" requires "graphql_union_types" to also be set', [ 'registered_post_type_object' => $post_type_object ] );

}
```

And that would allow an existing query to succeed, but get a helpful debug message: 

![CleanShot 2022-09-08 at 13 13 44](https://user-images.githubusercontent.com/1260765/189206797-5a0b1f28-6068-4bf5-a550-318defb5da58.png)

Reply:
> I'd hate to see a full on fatal error for the whole site when it's just a mistake affecting WPGraphQL.


My rationale for the fatal error is that since the schema API is solely for use by PHP developers, it'd be helpful to ensure mistakes get caught immediately instead of accidently slipping upstream. But i'm cool with the graceful approach too.

Do you also prefer`graphql_debug` over a throwing _GraphQL_ `Error`?

Reply:
It's hard to say. I think before plugins like Custom Post Type UI, Atlas Content Modeler, etc started allowing folks to do this stuff in a UI and not just in code, it's made it more of a gray area. 

If someone makes a mistake, I want to make sure it's easy to identify, but want to make sure it doesn't tank their existing client applications.

Reply:
I think more often, I'd lean on `graphql_debug` as it helps surface issues, but also allows folks to do their thing. 

Like here, we might end up not adding the new Union type to the schema, but we can still let existing queries execute. 

If they go looking for their new Type in the Schema docs or make a query for it, they'll know it's not registered properly and will have a good debug message for it, but exiting schema will remain unaffected.

Reply:
Probably case-by-case though, but feels like a generally good experience. Not so harsh that you can't complete your work, but also not so hidden that you can't troubleshoot easily.

Reply:
merged https://github.com/wp-graphql/wp-graphql/pull/2399/commits/263be0cc1aa5fc2296b99dce373d5924cb94ad9f

Reply:
@jasonbahl I think I'm missing something with the above approach.

`graphql_debug()` doesnt output anything on the query (I'm guessing since its happening _before_ the `register_graphql_*()` call. Neither does throwing a GraphQL error at that point in the lifecycle either.

Instead, the RootQuery field is simply not registered, and other queries dont show an issue either.

![image](https://user-images.githubusercontent.com/29322304/189498191-e2699765-bbdc-4e78-9ba7-e35dee5a7e8d.png)

I'm also not entirely sure how I could `graphql_debug()` within `TypeRegisty::register_type()` without losing the context (i.e. the post_type/tax that has a bad `graphql_` property).

> I think before plugins like Custom Post Type UI, Atlas Content Modeler, etc started allowing folks to do this stuff in a UI and not just in code, I used to prefer killing things like this and letting folks know right away. But now it's more of a gray area.
>
> If someone makes a mistake in ACF of CPT UI, this would fatal their site and they couldn't correct the mistake, which is harsh. I want to make sure it's easy to identify, but want to make sure it doesn't tank their existing client applications.

Perhaps the onus needs to be on those 3rd-party plugins that want to support this to make sure their UI requires and validates all the necessary fields?

Reply:
Well, the test here is trying to query something that is not in the schema (because we didn't add it to the schema) so it's getting a GraphQL error _before_ execution. 

If we did something like an introspection query the debug message would show.

Reply:
I definitely think we need to limit the exception/errors to a more narrow context than _any_
WordPress page load. No matter how much we recommend plugin developers follow some rule here, it will reflect real bad on WPGraphQL if one missing argument of the post type fatals your site. In code or via a GUI plugin. 

Id prefer a more graceful failure than a full on fatal. 

Possibly graphql_debug isn't the right call, but maybe as admin notice or something in addition to graphql_debug?  

Valid queries will output the debug message for developers, (we can at some point, surface the debug messages in GraphiQL better too) and the admin can notify folks of the error too, without killing the whole site.

Reply:
I'm open to other options for sure, but I'm opposed to fataling the whole site if a conditional array key/value is missing.

Reply:
We might be able to hook into one of the "validation" steps of GraphQL as well to throw a more clear error as to _why_ the field doesn't exist. 

In your test you register a post type, then make a query for it, even though we know it doesn't exist. 

If that query were made, perhaps at run time we can identify the field/type being queried for and if it matches a specific field/type that we have a specific error for we can surface it there, then for all other (valid) queries a debug message is surfaced

Reply:
Possibly something like one of these Validation Rules might work? 

src/Server/ValidationRules/RequireAuthentication.php

Reply:
> If we did something like an introspection query the debug message would show.

What I'm confused about is why it the debug message is showing [when you query `posts`](https://github.com/wp-graphql/wp-graphql/pull/2399#discussion_r966333256), but [not when I do](https://github.com/wp-graphql/wp-graphql/pull/2399#discussion_r967692003).